### PR TITLE
pulled out statice methods and type hinting #974

### PR DESCRIPTION
**Pull request by eric-s-s**
when i pulled out methods i needed to organize.  so i did:

1) main methods of main object
2) implementation details

this PR is almost entirely moving things around.  

because the act of changing position might make it hard to notice any changes, there are notes to myself on any additions/changes (very minor).  i'll be using those to comment every line that isn't exactly like the original (minus automatic chnages like removing `@staticmethod`).  and then those can be removed later.  

I'm new to type hinting and not sure the best way to go about it.  I tried making aliases for the different types at the top.  this looked nice in the code, but when i called up a docstring on my IDE, 
`EdgesToMatches` became `Dict[Tuple[int, int], List[List[Tuple[str, str]]]]`  which looks fugly.  should i just stick to very basic types?  should i set aliases to very basic types, like `EdgesToMatches = dict; PointsToFloats = dict` ?

### CODE REVIEW

File: axelrod/fingerprint.py [hunk 1]
```diff
@@ -1,60 +1,58 @@
...
 
-from axelrod import on_windows
+from axelrod import on_windows, Player, SpatialTournament
+from axelrod.actions import Action
 from axelrod.strategy_transformers import JossAnnTransformer, DualTransformer
 from axelrod.interaction_utils import (
     compute_final_score_per_turn, read_interactions_from_file)
 
 
 Point = namedtuple('Point', 'x y')
```
Comment (eric-s-s) [inline axelrod/fingerprint.py]: here are type hinting aliases.

File: axelrod/fingerprint.py [hunk 2]
```diff
@@ -1,60 +1,58 @@
...
 
-    return points
+EdgesToMatches = Dict[Edge, Matches]
+PointsToFloats = Dict[Point, float]
 
 
-class AshlockFingerprint():
-    def __init__(self, strategy, probe=axl.TitForTat):
+class AshlockFingerprint(object):
+    """Methodology for obtaining visual representation of a strategy's
```
Comment (eric-s-s) [inline axelrod/fingerprint.py]: added this summary

File: axelrod/fingerprint.py [hunk 3]
```diff
@@ -68,165 +66,17 @@ def __init__(self, strategy, probe=axl.TitForTat):
...
-            compute_final_score_per_turn(scores)[0]
-            for scores in interactions[edge]])
-            for edge in edges
-        ]
-        point_scores = dict(zip(points, edge_scores))
-        return point_scores
-
-    def fingerprint(self, turns=50, repetitions=10, step=0.01, processes=None,
-                    filename=None, in_memory=False, progress_bar=True):
+        self.data = None  # type: PointsToFloats
```
Comment (eric-s-s) [inline axelrod/fingerprint.py]: explicitly created all instance variable at __init__ and set to None with type hint.

File: axelrod/fingerprint.py [hunk 4]
```diff
@@ -240,34 +90,39 @@ def fingerprint(self, turns=50, repetitions=10, step=0.01, processes=None,
         turns : integer, optional
             The number of turns per match
         repetitions : integer, optional
-            The number of times the round robin should be repeated
+            The number of times the each match is repeated
         step : float, optional
+            0.0 < step <= 1.0
```
Comment (eric-s-s) [inline axelrod/fingerprint.py]: added sentence.

File: axelrod/fingerprint.py [hunk 5]
```diff
@@ -240,34 +90,39 @@ def fingerprint(self, turns=50, repetitions=10, step=0.01, processes=None,
...
         """
 
         if on_windows and (filename is None):  # pragma: no cover
             in_memory = True
         elif filename is None:
-            outputfile = NamedTemporaryFile(mode='w')
-            filename = outputfile.name
-
-        edges, tourn_players = self.construct_tournament_elements(
+            output_file = NamedTemporaryFile(mode='w')
```
Comment (eric-s-s) [inline axelrod/fingerprint.py]: variable typo cleanup.

File: axelrod/fingerprint.py [hunk 6]
```diff
@@ -337,13 +166,13 @@ def plot(self, col_map='seismic', interpolation='none', title=None,
         figure : matplotlib figure
             A heat plot of the results of the spatial tournament
         """
-        size = int((1 / self.step) // 1) + 1
-        plotting_data = self.reshape_data(self.data, self.points, size)
+        size = int(1 / self.step) + 1
```
Comment (eric-s-s) [inline axelrod/fingerprint.py]: changed this line of code to remove repetition.

File: axelrod/fingerprint.py [hunk 7]
```diff
@@ -361,3 +190,236 @@ def plot(self, col_map='seismic', interpolation='none', title=None,
...
+
+    def construct_tournament_elements(
+        self, step: float, progress_bar: bool = True
+    ) -> Tuple[EdgeList, PlayerList]:
+        """Build the elements required for a spatial tournament
+
+        Parameters
+        ----------
+        step : float
+            0.0 < step <= 1.0.
```
Comment (eric-s-s) [inline axelrod/fingerprint.py]: added this line of docstring.

File: axelrod/fingerprint.py [hunk 8]
```diff
@@ -361,3 +190,236 @@ def plot(self, col_map='seismic', interpolation='none', title=None,
...
+        Points with coordinates that will be closer together.
+    progress_bar : bool
+        Whether or not to create a progress bar which will be updated
+
+    Returns
+    ----------
+    points : list
+        of Point objects with coordinates (x, y)
+    """
+    num = int(1 / step) + 1
```
Comment (eric-s-s) [inline axelrod/fingerprint.py]: changed this from `num = int((1 / step) // 1) + 1` to remove repetition.

File: axelrod/fingerprint.py [hunk 9]
```diff
@@ -361,3 +190,236 @@ def plot(self, col_map='seismic', interpolation='none', title=None,
...
+    """Creates a JossAnn probe player that matches the Point.
+
+    If the coordinates of point sums to more than 1 the parameters are
+    flipped and subtracted from 1 to give meaningful probabilities. We also
+    use the Dual of the probe. This is outlined further in [Ashlock2010]_.
+
+    Parameters
+    ----------
+    point : Point
+    probe : class or Player
```
Comment (eric-s-s) [inline axelrod/fingerprint.py]: changed this to be class or Player instead of just class

File: axelrod/tests/unit/test_fingerprint.py [hunk 10]
```diff
@@ -11,6 +28,141 @@
     matplotlib_installed = False
 
 
+class TestFingerPrintHelpers(unittest.TestCase):
+
+    def test_create_points_progress_bar_true(self):
```
Comment (eric-s-s) [inline axelrod/tests/unit/test_fingerprint.py]: new tests for create_points.  I had expected create_points(0.3) to return points in [0.0, 0.3, 0.9]  so i thought it would be good to explicitly state that is not so.

File: axelrod/tests/unit/test_fingerprint.py [hunk 11]
```diff
@@ -11,6 +28,141 @@
...
+        points = [(0.0, 0.0), (0.0, 0.5), (0.0, 1.0),
+                  (0.5, 0.0), (0.5, 0.5), (0.5, 1.0),
+                  (1.0, 0.0), (1.0, 0.5), (1.0, 1.0)]
+        expected_edges = [(0, 1), (0, 2), (0, 3),
+                          (0, 4), (0, 5), (0, 6),
+                          (0, 7), (0, 8), (0, 9)]
+        edges = create_edges(points, progress_bar=False)
+        self.assertEqual(edges, expected_edges)
+
+    def test_create_edges_progress_bar_true(self):
```
Comment (eric-s-s) [inline axelrod/tests/unit/test_fingerprint.py]: new test.

File: axelrod/tests/unit/test_fingerprint.py [hunk 12]
```diff
@@ -11,6 +28,141 @@
...
+        # x + y > 1
+        ja = create_jossann((.5, .6), probe)
+        self.assertEqual(str(ja), "Dual Joss-Ann Random: 0.1: (0.5, 0.4)")
+
+    def test_create_probes_progress_bar_true(self):
+        nine_points = create_points(0.5, progress_bar=False)
+        probes = create_probes(axl.TitForTat, nine_points, progress_bar=True)
+        self.assertEqual(len(probes), 9)
+
+    def test_create_probes(self):
```
Comment (eric-s-s) [inline axelrod/tests/unit/test_fingerprint.py]: test changed from original. not by much, just shorter list, i think.

File: axelrod/tests/unit/test_fingerprint.py [hunk 13]
```diff
@@ -11,6 +28,141 @@
...
+        expected_probe_str = ['Joss-Ann Tit For Tat: (0.0, 0.0)',
+                              'Dual Joss-Ann Tit For Tat: (1.0, 0.0)',
+                              'Dual Joss-Ann Tit For Tat: (0.0, 1.0)',
+                              'Dual Joss-Ann Tit For Tat: (0.0, 0.0)']
+        points = create_points(1.0, progress_bar=False)
+        probes = create_probes(axl.TitForTat, points, progress_bar=False)
+        self.assertEqual([str(probe) for probe in probes], expected_probe_str)
+
+    def test_generate_data(self):
```
Comment (eric-s-s) [inline axelrod/tests/unit/test_fingerprint.py]: new test.

File: axelrod/tests/unit/test_fingerprint.py [hunk 14]
```diff
@@ -1,7 +1,24 @@
+"""
+added or modified tests
+
```
Comment (eric-s-s) [inline axelrod/tests/unit/test_fingerprint.py]: the lion's share of tests are simply in different places, and the ones not in TestAshlockFingerprint test their original data, but it's in the test.

anything not EXACTLY like it was is marked.

File: axelrod/tests/unit/test_fingerprint.py [hunk 15]
```diff
@@ -49,57 +201,18 @@ def test_init_with_instance(self):
...
-        af = AshlockFingerprint(self.strategy, self.probe)
-        edges = af.create_edges(self.expected_points, progress_bar=False)
+        edges, tournament_players = af.construct_tournament_elements(
+            0.5, progress_bar=False)
         self.assertEqual(edges, self.expected_edges)
+        self.assertEqual(len(tournament_players), 10)
+        self.assertEqual(tournament_players[0].__class__, af.strategy)
 
-    def test_construct_tournament_elemets(self):
+    def test_construct_tournament_elements_progress_bar_true(self):
```
Comment (eric-s-s) [inline axelrod/tests/unit/test_fingerprint.py]: new test.  copies old test exactly, which is below.

File: axelrod/tests/unit/test_fingerprint.py [hunk 16]
```diff
@@ -119,6 +232,19 @@ def test_fingerprint_with_filename(self):
             data = out.read()
             self.assertEqual(len(data.split("\n")), 10)
 
+    def test_fingerprint_with_filename_none_and_in_memory_false(self):
```
Comment (eric-s-s) [inline axelrod/tests/unit/test_fingerprint.py]: this test is new.  it just failed (since i use windows, it had different behavior on my machine) .

File: axelrod/tests/unit/test_fingerprint.py [hunk 17]
```diff
@@ -119,6 +232,19 @@ def test_fingerprint_with_filename(self):
...
+        af = AshlockFingerprint(self.strategy, self.probe)
+        af.fingerprint(turns=1, repetitions=1, step=0.5, progress_bar=False,
+                       filename=None, in_memory=False)
+
+        edge_keys = sorted(list(af.interactions.keys()))
+        coord_keys = sorted(list(af.data.keys()))
+        self.assertEqual(af.step, 0.5)
+        self.assertEqual(af.spatial_tournament.interactions_dict,
```
Comment (eric-s-s) [inline axelrod/tests/unit/test_fingerprint.py]: it failed right here.  should this have happened, or is this an issue?

File: axelrod/tests/unit/test_fingerprint.py [hunk 18]
```diff
@@ -325,21 +408,21 @@ def test_pair_fingerprints(self, strategy_pair):
         """
         strategy, probe = strategy_pair
         af = AshlockFingerprint(strategy, probe)
-        data = af.fingerprint(turns=2, repetitions=2, step=0.5,
+        data = af.fingerprint(turns=2, repetitions=2, step=1.0,
```
Comment (eric-s-s) [inline axelrod/tests/unit/test_fingerprint.py]: chnaged these to 1.0 to cut my testing time in half.

File: axelrod/tests/unit/test_fingerprint.py [hunk 19]
```diff
@@ -119,6 +232,14 @@ def test_fingerprint_with_filename(self):
             data = out.read()
             self.assertEqual(len(data.split("\n")), 10)
 
+    def test_fingerprint_with_filename_none_and_in_memory_false(self):
```
Comment (eric-s-s) [inline axelrod/tests/unit/test_fingerprint.py]: new test.

File: axelrod/tests/unit/test_fingerprint.py [hunk 20]
```diff
@@ -119,6 +233,27 @@ def test_fingerprint_with_filename(self):
             data = out.read()
             self.assertEqual(len(data.split("\n")), 10)
 
+    @unittest.skipIf(axl.on_windows,
```
Comment (eric-s-s) [inline axelrod/tests/unit/test_fingerprint.py]: new tests.  hopefully these work.  submitting blind as on windows machine.  off to work. fix later if needed.

File: axelrod/fingerprint.py [hunk 21]
```diff
@@ -242,32 +92,37 @@ def fingerprint(self, turns=50, repetitions=10, step=0.01, processes=None,
...
         Returns
         ----------
-        self.data : dictionary
+        self.data : dictionary(Point, float)
             A dictionary where the keys are coordinates of the form (x, y) and
             the values are the mean score for the corresponding interactions.
         """
 
-        if on_windows and (filename is None):  # pragma: no cover
+        if axl.on_windows and (filename is None):  # pragma: no cover
```
Comment (eric-s-s) [inline axelrod/fingerprint.py]: did a direct call here (instead of top_level import) so that `unittest.mock.patch` will work in testing.  if test is a bust, will revert.

**Comment by eric-s-s** [comment]
I looked at the test failure.  does this have anything to do with what I wrote?  it doesn't look like there should be any connection. is there a way to run the tests again without submitting a PR?

**Comment by drvinceknight** [comment]
> is there a way to run the tests again without submitting a PR?

It's a hypothesis thing. I've set the build going again.

File: axelrod/tests/unit/test_fingerprint.py (hunk 18 above)
Comment (drvinceknight) [inline axelrod/tests/unit/test_fingerprint.py]: This is an edge case though. Keep it at `0.5` please so that it does test all point creations.

**Comment by drvinceknight** [comment]
I'll get to this properly tomorrow (at a meetup this evening :) :+1:)

File: axelrod/fingerprint.py [hunk 22]
```diff
@@ -1,60 +1,58 @@
+"""
```
Comment (drvinceknight) [inline axelrod/fingerprint.py]: Can we remove this please (it will become a "lie" if further changes are made that don't get recorded here). We can see the changes via the commits (and their messages can be used to in effect describe the same thing you're doing here).

File: axelrod/fingerprint.py [hunk 23]
```diff
@@ -1,60 +1,58 @@
...
+from axelrod import Player, SpatialTournament
+from axelrod.actions import Action
 from axelrod.strategy_transformers import JossAnnTransformer, DualTransformer
 from axelrod.interaction_utils import (
     compute_final_score_per_turn, read_interactions_from_file)
 
 
 Point = namedtuple('Point', 'x y')
 
+PlayerOrStrategy = Union[Player, type]
```
Comment (drvinceknight) [inline axelrod/fingerprint.py]: I personally do not like type aliases for type hinting. They aren't helpful when accessing the docstring/syntax of a function from elsewhere. I suggest these just get written verbosely throughout.

File: axelrod/fingerprint.py (hunk 5 above)
Comment (drvinceknight) [inline axelrod/fingerprint.py]: Where's the typo?

File: axelrod/fingerprint.py [hunk 24]
```diff
@@ -327,7 +156,7 @@ def plot(self, col_map='seismic', interpolation='none', title=None,
             http://matplotlib.org/examples/images_contours_and_fields/interpolation_methods.html
         title : str, optional
             A title for the plot
-        colorbar : bool, optional
+        color_bar : bool, optional
```
Comment (drvinceknight) [inline axelrod/fingerprint.py]: Change this back to `colorbar`. This is inline with `matplotlib`'s variable/method name (which is why the variable was called this to begin with) and this also introduces backwards incompatibility. 

Before I look through further please revert any variable names/function names (that are not relevant to this refactor) back to what they were. Having to re explain the reasoning behind every previous choice is just going to be too time consuming. Once you feel this PR is a simple refactor of the methods to functions let me know and I'll review.

File: axelrod/fingerprint.py (hunk 22 above)
Comment (meatballs) [inline axelrod/fingerprint.py]: Agreed. There is no need for this. That's what version control is for.

File: axelrod/fingerprint.py (hunk 5 above)
Comment (eric-s-s) [inline axelrod/fingerprint.py]: the linter complained that `outputfile` and `tourn_players` were typos, but `output_file` and `tournament_players` were ok. it seemed reasonable.

Comment (meatballs) [inline axelrod/fingerprint.py]: Are you sure that wasn't a spell checker complaining? There's no linting problem here for me

File: axelrod/fingerprint.py (hunk 24 above)
Comment (meatballs) [inline axelrod/fingerprint.py]: Agreed. What problem is this fixing?

File: axelrod/fingerprint.py (hunk 22 above)
Comment (eric-s-s) [inline axelrod/fingerprint.py]: ahh.  sorry.  i should have put this top comment here

>because the act of changing position might make it hard to notice any changes, there are notes to myself on any additions/changes (very minor). i'll be using those to comment every line that isn't exactly like the original (minus automatic chnages like removing @staticmethod). and then those can be removed later.

these were those comments i was going to remove.

File: axelrod/fingerprint.py (hunk 24 above)
Comment (eric-s-s) [inline axelrod/fingerprint.py]: no problem.

File: axelrod/fingerprint.py (hunk 21 above)
Comment (eric-s-s) [inline axelrod/fingerprint.py]: so, i just want to confirm.  keep this and the relevant tests to go with it, or ditch it?  (all the other changes will be gone in a moment.)

File: axelrod/fingerprint.py (hunk 23 above)
Comment (eric-s-s) [inline axelrod/fingerprint.py]: thank you.  that answered my question nicely.

File: axelrod/fingerprint.py (hunk 21 above)
Comment (drvinceknight) [inline axelrod/fingerprint.py]: This sounds like a good change: it's inline with the issue (moving static methods to functions and better isolation of tests) and is not a stylistic change :) :+1: 

It also make no functionality change.

**Comment by eric-s-s** [comment]
wait a moment, please.  missed a few tests.

**Comment by eric-s-s** [comment]
i'm going to close this.  easier to just start from scratch.
